lib/service_actor/type_checkable.rb

```diff
@@ -26,18 +49,44 @@ def _call
 
     private
 
-    def check_type_definitions(definitions, kind:)
+    def check_type_definitions(definitions, kind:) # rubocop:disable Metrics/MethodLength
       definitions.each do |key, options|
         type_definition = options[:type] || next
         value = result[key] || next
 
-        types = types_for_definition(type_definition)
+        base_arguments = {
+          kind: kind,
+          input_key: key,
+          service_name: self.class,
+          actual_type_name: value.class
+        }
+
+        types, message = define_types_with(type_definition)
+
         next if types.any? { |type| value.is_a?(type) }
 
-        raise ServiceActor::ArgumentError,
-              "#{kind} #{key} on #{self.class} must be of type " \
-              "#{types.join(', ')} but was #{value.class}"
+        raise_error_with(
+          message,
+          **base_arguments,
+          expected_type_names: types.join(", "),
+        )
+      end
+    end
+
+    def define_types_with(type_definition)
+      message = DEFAULT_MESSAGE
+
+      if type_definition.is_a?(Hash) # advanced mode
+        type_definition, message =
+          type_definition.values_at(:is, :message)
       end
+
+      types = types_for_definition(type_definition)
+
+      [
+        types,
+        message
+      ]
```

Comment:
I think it could make it a tiny bit clearer if we move the default in an `else`:

```suggestion
      if type_definition.is_a?(Hash)
        type_definition, message =
          type_definition.values_at(:is, :message)
      else
        message = DEFAULT_MESSAGE
      end

      types = types_for_definition(type_definition)

      [types, message]
```